counterpartylib/lib/api.py

```diff
@@ -382,21 +384,34 @@ def run(self):
 
 class APIServer(threading.Thread):
     """Handle JSON-RPC API calls."""
-    def __init__(self):
+    def __init__(self, db=None):
+        self.db = db
         self.is_ready = False
+        self.shutdown_secret = binascii.hexlify(os.urandom(32)).decode('ascii')
         threading.Thread.__init__(self)
         self.stop_event = threading.Event()
 
+        self.app = None
+
     def stop(self):
+        util.api('shutdown', {'secret': self.shutdown_secret})
         self.join()
         self.stop_event.set()
 
     def run(self):
         logger.info('Starting API Server.')
-        db = database.get_connection(read_only=True, integrity_check=False)
+        self.db = self.db or database.get_connection(read_only=True, integrity_check=False)
         app = flask.Flask(__name__)
         auth = HTTPBasicAuth()
 
+        @dispatcher.add_method
+        def shutdown(secret):
+            assert self.shutdown_secret and secret == self.shutdown_secret
```

Comment:
Good catch!! This likely exists in another branch as well!!

I think this may have come from another branch I depended upon that has not been merged into dev yet. @rubensayshi do you know more about this?